doc/specs/#476 - Package Pinning.md

```diff
@@ -0,0 +1,205 @@
+---
+author: Yao Sun @yao-msft
+created on: 2022-10-12
+last updated: 2022-10-12
+issue id: 476
+---
+
+# Package Pinning
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality and high level implementation design of Package Pinning feature.
+
+## Inspiration
+
+This is inspired by functionalities in other package managers, as well as community feedback.
+- Packages may introduce breaking changes that users may not want integrate into their workflow quite yet.
+- Packages may update themselves so that it will be duplicate effort for winget to try to update them.
+- User may want to maintain some of the packages through other channels outside of winget, or prefer one source over others within winget.
+- User may want some of the packages to stay in some major versions but allow minor version changes during upgrade.
+
+## Solution Design
+
+#### Package Pinning types
+To achieve goals listed above, winget will support 3 types of Package Pinning:
+- **Blocking:** The package is blocked from `winget upgrade --all` or `winget upgrade <specific package>`, user has to unblock the package to let winget perform upgrade.
+- **Pinning:** The package is excluded from `winget upgrade --all` but allowed in `winget upgrade <specific package>`, a new argument `--include-pinned` will be introduced to let `winget upgrade --all` to include pinned packages.
+- **Gating:** The package is pinned to specific version(s). For example, if a package is pinned to version `1.2.*`, any version between `1.2.0` to `1.2.<anything>` is considered valid.
+
+To allow user override, `--force` can be used with `winget upgrade <specific package>` to override some of the pinning created above.
```

Comment:
I thought we wanted to move away from `--force` in favor of more specific flags?

Reply:
If we go with "--force", we should be explicit about what messaging the user receives in terms of the current state of a pin, and what the new state should be.

Reply:
> This isn't security related. It was my understanding that the specificity was around security related concerns to prevent users from just `--force`-ing their way through. If it isn't a security concern, I think `--force` is appropriate?

Yes, I think that's what we intended some time ago. For security related, we should be specific to give user precise control. For non-security related, we prefer --force to avoid too many switches unless it's semantically special that deserves a separate switch.